# Numpy - Domača naloga

Uvozimo modul numpy:

In [3]:
import numpy as np

### 1.Create a null vector of size 10

In [10]:
vec_1 = np.zeros(10)
print(f"Odg: {vec_1}")

Odg: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]`

### 2.Create a null vector of size 10 but the fifth value which is 1

In [14]:
vec_2 = np.zeros(10)
vec_2[4] = 1
print(f"Odg: {vec_2}")

Odg: [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]


Odg: `[0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]`

### 3.Create a vector with values ranging from 10 to 49

In [22]:
vec_3 = np.linspace(10,49,40)
print(vec_3)

[10. 11. 12. 13. 14. 15. 16. 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27.
 28. 29. 30. 31. 32. 33. 34. 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45.
 46. 47. 48. 49.]


Odg: `[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49]`

### 4.Reverse a vector (first element becomes last)

In [29]:
Z = np.arange(50)
vec_4 = np.flip(Z)
print(vec_4)

[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26
 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2
  1  0]


Odg: `[49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33 32 31 30 29 28 27 26 25 24 23 22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2 1  0]`

### 5.Create a 3x3 matrix with values ranging from 0 to 8

**Pomoč**: metoda .reshape() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.reshape.html)

In [41]:
vec_5 = np.linspace(0,8,9, dtype=int)
# print(vec_5)
mtrx_5 = vec_5.reshape((3,3))
print(mtrx_5)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


Odg: 

`[[0 1 2]
 [3 4 5]
 [6 7 8]]`

### 6.Find indices of non-zero elements from [1,2,0,0,4,0]

**Pomoč:** Google, Stack Overflow, Numpy dokumentacija

In [4]:
a = [1,2,0,0,4,0]
vec_6 = np.nonzero(a)
print(vec_6)


(array([0, 1, 4], dtype=int64),)


Odg: `(array([0, 1, 4], dtype=int64),)`

### 7.Create a 3x3 identity matrix

In [8]:
mat_7 = np.identity(3)
print(mat_7)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


Odg: 

`[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]`

### 8.Create a 3x3x3 array with random values

In [11]:
mat_8 = np.random.rand(3,3,3)
print(mat_8)

[[[0.48565488 0.60180236 0.93790791]
  [0.37804279 0.78396235 0.84294855]
  [0.12994277 0.97594767 0.03072848]]

 [[0.63339962 0.1721612  0.39663588]
  [0.54204868 0.72971805 0.17313314]
  [0.73897934 0.32645245 0.93706511]]

 [[0.07006164 0.52864514 0.56955456]
  [0.84235618 0.57752829 0.91100464]
  [0.31040416 0.91754211 0.96762984]]]


Odgovor npr.: 
    
`[[[0.61712056 0.38943629 0.16277319]
  [0.19481511 0.12725415 0.6717974 ]
  [0.15340477 0.78713636 0.52864134]]

 [[0.91153827 0.7756351  0.88600077]
  [0.03509814 0.79676837 0.39167056]
  [0.55410251 0.67583417 0.64169036]]

 [[0.46862042 0.05954143 0.42512482]
  [0.34746791 0.97653267 0.0024134 ]
  [0.53522577 0.37192358 0.29601563]]]`

### 9.Create a 10x10 array with random values and find the minimum and maximum values

In [17]:
arr_9 = np.random.rand(10,10)
min_value = np.min(arr_9)
max_value = np.max(arr_9)
print(arr_9)
print(min_value)
print(max_value)

[[0.83090633 0.10773515 0.92244901 0.64480139 0.159124   0.07686022
  0.8735612  0.87098351 0.08415661 0.93243291]
 [0.79151466 0.8389693  0.15236586 0.72276824 0.21816539 0.0228067
  0.56191173 0.80518427 0.05440851 0.10433603]
 [0.08464702 0.84988167 0.89596023 0.1693317  0.15743839 0.60678239
  0.39253954 0.07470793 0.35444044 0.04158501]
 [0.06631206 0.49132158 0.55803059 0.24135398 0.71015917 0.43747944
  0.61794384 0.4656985  0.51500748 0.96499204]
 [0.04093224 0.68723894 0.51783083 0.28555204 0.34518223 0.89672639
  0.96220906 0.75493747 0.16059004 0.44645009]
 [0.22285937 0.23756076 0.26890107 0.22474426 0.17486584 0.04125721
  0.28826664 0.64584474 0.59888625 0.3223376 ]
 [0.26632896 0.80764756 0.06196575 0.59808491 0.89665839 0.83698521
  0.3909278  0.06973753 0.11880573 0.63591746]
 [0.12373747 0.88811168 0.93243885 0.23656014 0.2922914  0.20670518
  0.9092141  0.41185635 0.92021012 0.04213269]
 [0.21212519 0.96538879 0.26982853 0.16860966 0.99187857 0.06091613
  0.83809472 

Odg: `Zmin= 0.004874889654358006 Zmax= 0.992254173881756`

### 10.Create a 2d array (10x10) with 1 on the border and 0 inside

In [25]:
arr_10 = np.zeros((10,10))
arr_10[0,:] = 1
arr_10[-1,:] = 1
arr_10[:,0] = 1
arr_10[:,-1] = 1
print(arr_10)

[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]


Odg: 
    
`[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]`

### 11.What is the result of the following expression?

<div class="alert alert-block alert-info">Result of zero multiplied by nan is nan!</div>

In [26]:
0 * np.nan

nan

<div class="alert alert-block alert-info">Do not compare or calculate with nan values!</div>

In [27]:
np.nan == np.nan

False

In [28]:
np.inf > np.nan

False

In [29]:
np.nan - np.nan

nan

<div class="alert alert-block alert-info"> Do not compare floating points. Use interval abs(a-b) smaller than Eps. </div>

In [30]:
0.3 == 3 * 0.1 

False

### 12.Create a 8x8 matrix and fill it with a checkerboard pattern

In [79]:
mat_12 = np.indices((8,8)).sum(axis=0) % 2
print(mat_12)


[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


Odg: 

`[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]`

### 13.Given a 1D array, negate all elements which are between 3 and 8, in place.

In [109]:
Z = np.arange(11)

# primer 1
mask = (Z > 3) & (Z < 9)
Z[mask] *= -1
print(Z)

# primer 2
# mask = np.where((Z > 3) & (Z < 9),-1,1)
# result = Z * mask
# print(result)


[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]


Odg: `[ 0  1  2  3 -4 -5 -6 -7 -8  9 10]`

### 14.Create a vector of size 10 with values ranging from 0 to 1, both excluded

In [130]:
vec_14 = np.linspace(0,1,12)[1:-1]
print(vec_14)

[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455
 0.63636364 0.72727273 0.81818182 0.90909091]


Odg: `[0.09090909 0.18181818 0.27272727 0.36363636 0.45454545 0.54545455 0.63636364 0.72727273 0.81818182 0.90909091]`

### 15.Create a random vector of size 10 and sort it

In [134]:
vec_15 = np.sort(np.random.rand(10))
print(vec_15)

[0.03847999 0.13125731 0.16823507 0.20708885 0.31625267 0.36961721
 0.37000394 0.55674482 0.59831067 0.63601582]


Odg. npr.: `[0.08755811 0.17786018 0.45041482 0.48746541 0.68379823 0.70209576 0.77273884 0.86106446 0.93083351 0.98349603]`

### 16. Consider two random array A anb B, check if they are equal

In [135]:
A = np.random.randint(0,2,5)
print(A)

[1 0 1 0 0]


In [136]:
B = np.random.randint(0,2,5)
print(B)

[1 0 0 0 0]


In [137]:
np.array_equal(A, B)

False

### 17.Make an array immutable (read-only)

In [138]:
Z = np.zeros(10)
Z.flags.writeable = False
Z[0] = 1

ValueError: assignment destination is read-only

### 18.Create random vector of size 10 and replace the maximum value by 0

In [169]:
vec_18 = np.random.rand(10)
vec_18[np.argmax(vec_18)] = 0
print(vec_18)

[0.7792805  0.57949462 0.49289705 0.         0.00793193 0.48029022
 0.25156214 0.25761526 0.68552748 0.80913388]


### 19.How to find the closest value (to a given scalar -v) in an array -Z?

In [183]:
Z = np.arange(100)
v = np.random.uniform(0,100)
print(v)
print(Z[abs(Z-v).argmin()])


12.990631351070746
13


### 20.How to convert an integer (32 bits)  array into a float (32 bits) in place?

In [185]:
Z = np.arange(10, dtype=np.int32)
Z.astype(dtype=np.float32)

array([0., 1., 2., 3., 4., 5., 6., 7., 8., 9.], dtype=float32)

Odg: `[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]`

### 21.How to find the most frequent value in an array?

**Pomoč:** metoda .bincount() -> [Dokumentacija](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html?highlight=bincount#numpy.bincount)

Histogram razporeditve po "binih" (koših) -> [0,1) [1,2) [2,3) ...

In [198]:
Z = np.random.randint(0,10,50)
print(Z)
histogram = np.bincount(Z)
print(histogram)
indices = histogram.argmax()
print(f"Most frequent value is {indices}") # glede na definicije binov index kar ustreza vrednosti

[8 3 4 1 5 8 3 4 0 6 3 8 2 3 0 4 0 1 4 2 5 4 2 6 7 0 9 0 2 8 9 6 3 7 4 5 0
 8 1 8 1 3 9 1 5 5 5 1 3 0]
[7 6 4 7 6 6 3 2 6 3]
Most frequent value is 0
